src/Aspire.Hosting.Azure/ExistingAzureResourceExtensions.cs

```diff
@@ -0,0 +1,80 @@
+    /// Determines if the resource is an existing resource.
+    /// </summary>
+    /// <param name="resource">The resource to check.</param>
+    /// <returns>True if the resource is an existing resource, otherwise false.</returns>
+    public static bool IsExisting(this IResource resource)
+    {
+        ArgumentNullException.ThrowIfNull(resource);
+
+        return resource.Annotations.OfType<ExistingAzureResourceAnnotation>().Any();
+    }
+
+    /// <summary>
+    /// Marks the resource as an existing resource when the application is running.
+    /// </summary>
+    /// <typeparam name="T">The type of the resource.</typeparam>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the existing resource.</param>
+    /// <returns>The resource builder with the existing resource annotation added.</returns>
+    public static IResourceBuilder<T> RunAsExisting<T>(this IResourceBuilder<T> builder, IResourceBuilder<ParameterResource> name)
+        where T : IAzureResource
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        if (!builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            builder.Resource.Annotations.Add(new ExistingAzureResourceAnnotation(name));
+        }
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Marks the resource as an existing resource when the application is deployed.
+    /// </summary>
+    /// <typeparam name="T">The type of the resource.</typeparam>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="name">The name of the existing resource.</param>
+    /// <returns>The resource builder with the existing resource annotation added.</returns>
+    public static IResourceBuilder<T> PublishAsExisting<T>(this IResourceBuilder<T> builder, IResourceBuilder<ParameterResource> name)
+        where T : IAzureResource
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        if (builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            builder.Resource.Annotations.Add(new ExistingAzureResourceAnnotation(name));
+        }
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Gets the <see cref="ExistingAzureResourceAnnotation" /> if the resource is an existing resource.
+    /// </summary>
+    /// <param name="resource">The resource to check.</param>
+    /// <param name="existingAzureResourceAnnotation">The existing resource annotation if the resource is an existing resource.</param>
+    /// <returns>True if the resource is an existing resource, otherwise false.</returns>
+    public static bool TryGetExistingAzureResourceAnnotation(this IResource resource, [NotNullWhen(true)] out ExistingAzureResourceAnnotation? existingAzureResourceAnnotation)
+    {
+        ArgumentNullException.ThrowIfNull(resource);
+
+        existingAzureResourceAnnotation = resource.Annotations.OfType<ExistingAzureResourceAnnotation>().FirstOrDefault();
```

Comment:
For annotaitons, we usually have either:

* respect multiple (not applicable here)
* expect single
* Last one wins

https://github.com/dotnet/aspire/blob/706bbbf4d12957ac0ada81f4247f65c2e8b54c88/src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs#L277-L279

I don't think using the "first" one is typical.

Reply:
Good point -- the intent here was to follow the expect single pattern.